examples/getting_started/tools_and_agents.ipynb

```diff
@@ -0,0 +1,586 @@
+    "            for element in soup.find_all(tag):\n",
+    "                element.decompose()\n",
+    "        main_content = (\n",
+    "            soup.find(\"main\")\n",
+    "            or soup.find(\"article\")\n",
+    "            or soup.find(\"div\", class_=re.compile(\"content|main\"))\n",
+    "        )\n",
+    "        if main_content:\n",
+    "            text = main_content.get_text(separator=\"\\n\", strip=True)\n",
+    "        else:\n",
+    "            text = soup.get_text(separator=\"\\n\", strip=True)\n",
+    "        lines = (line.strip() for line in text.splitlines())\n",
+    "        return \"\\n\".join(line for line in lines if line)\n",
+    "    except Exception as e:\n",
+    "        return f\"{type(e)}: Failed to extract content from URL {url}\""
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This `extract_content` function is a Tool that takes a URL as input and returns the main content of the webpage as a string. It uses [BeautifulSoup](https://www.crummy.com/software/BeautifulSoup/) to parse the HTML and extract the relevant text content.\n",
+    "\n",
+    "When this function is passed to a Mirascope `call` decorator, it is automatically converted into a `BaseTool` class. This conversion process generates a schema that the LLM API uses to understand and interact with the tool.\n",
+    "\n",
+    "Let's take a look at what this schema looks like:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {
+    "ExecuteTime": {
+     "end_time": "2024-09-12T01:39:50.790282Z",
+     "start_time": "2024-09-12T01:39:50.757565Z"
+    }
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "{'function': {'name': 'extract_content', 'description': 'Extract the main content from a webpage.\\n\\nArgs:\\n    url: The URL of the webpage to extract the content from.\\n\\nReturns:\\n    The extracted content as a string.', 'parameters': {'properties': {'url': {'description': 'The URL of the webpage to extract the content from.', 'type': 'string'}}, 'required': ['url'], 'type': 'object'}}, 'type': 'function'}\n"
+     ]
+    }
+   ],
+   "source": [
+    "from mirascope.core.openai import OpenAITool\n",
+    "\n",
+    "tool_type = OpenAITool.type_from_fn(extract_content)\n",
+    "print(tool_type.tool_schema())"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This schema provides the LLM with information about the tool's name, description, and expected input parameters. The LLM uses this information to determine when and how to use the tool during its reasoning process.\n",
+    "\n",
+    "For more details on implementing and using Tools in Mirascope, see the [Tools documentation](https://docs.mirascope.io/latest/learn/tools/)."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 5. Creating Custom Tools\n",
```

Comment:
nit: after reading this again, I'm wondering if this is the right. Maybe it would be better to write something like "Tools With Access To Agent State"?